content/blog/2022/2022-11-21-building-csv-importer-with-nodejs.md

```diff
@@ -0,0 +1,953 @@
+
+```bash
+npm init
+```
+
+Next, we need to install a few libraries as dependencies:
+
+```bash
+npm install express cors multer pg sequelize fast-csv json2csv
+```
+
+Now, let’s understand the various uses of the installed dependencies:
+
+* **Express** - This will be used to configure the Express server for REST API.
+* **Cors** - This is used for CORS(Cross-Origin Resource Sharing) configuration between the backend and the frontend server.
+* **Multer** - It is a NodeJs middleware used for handling `multipart/form-data`, which is primarily used for uploading files.
+* **Pg** - It is a non-blocking PostgreSQL client for NodeJs.
+* **Sequelize** - This is a modern TypeScript and Node.js ORM for various databases like Oracle, Postgres, MySQL, MariaDB, SQLite and SQL Server.
+* **Fast-csv** - This library is used for parsing and formatting CSVs or any other delimited value file in NodeJs.
+* **Json2csv** - This library is used to convert JSON into CSV with column titles and proper line endings.
+
+Now, we will create a server folder and add all our code within that directory.
+
+Next, we need to define the frontend React client. So we will create another directory to host our frontend.
+
+```bash
+npx create-react-app client
+```
+
+This will bootstrap the React code under the client folder. We will first implement the backend part and then we will come back to the frontend side.
+
+## Configure a PostgreSQL Database
+
+We have the base setup for our implementation ready. So, let’s host an instance of PostgreSQL and configure our backend server to connect with that DB. We can quickly spin a PostgreSQL instance by creating a docker-compose file:
+
+```yaml
+version: '3.1'
+
+services:
+  db:
+    image: postgres
+    restart: always
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: Welcome123
+      POSTGRES_DB: csvdb
+```
+
+We can run this by executing the following command:
+
+```bash
+docker-compose up
+```
+
+This will host a Postgres instance locally. Now we can switch to our code and create a `config` under the `server` directory. Then we need to define the database connection details:
+
+```javascript
+const HOST = "localhost";
+const USER = "postgres";
+const PASSWORD = "Welcome123";
+const DB = "csvdb";
+const dialect = "postgres";
+const pool = {
+    max: 5,
+    min: 0,
+    acquire: 30000,
+    idle: 10000
+};
+
+export default {
+    HOST,
+    USER,
+    PASSWORD,
+    DB,
+    dialect,
+    pool
+};
+```
+
+The first five details are specific to the PostgreSQL driver. We have also defined an optional parameter to configure the connection `pool` for Sequelize.
+
+## Defining the Data Model
+
+Next, we can initialize a data model for Sequelize. Sequelize is an ORM, so we need to define a model for the `Employee` data that we want to store in our database. We can create a `models` folder and define a model:
+
+```javascript
+import Sequelize from 'sequelize';
```

Comment:
Again, what is the name of this file? Can you please go through all code examples and mention the file names in the sentence before each code example? That makes it easier for readers to recognize the code in the example project.

Reply:
Updated the exact file name.